### PR TITLE
Dev loll

### PR DESCRIPTION
**Pull request by lollcat**
Solves #61 
Add specification of alpha in fab loss (both with and without buffer)

### CODE REVIEW

File: experiments/aldp/train.py [hunk 1]
```diff
@@ -15,7 +15,7 @@
 from fab.utils.numerical import effective_sample_size
 from fab.utils.replay_buffer import ReplayBuffer
 from fab.utils.prioritised_replay_buffer import PrioritisedReplayBuffer
-from fab.core import P_SQ_OVER_Q_TARGET_LOSSES
+from fab.core import ALPHA_DIV_TARGET_LOSSES
```
Comment (lollcat) [inline experiments/aldp/train.py:18]: I'm not super confident in editing the aldp file - could you please check I've done it correctly?

File: fab/sampling_methods/base.py [hunk 2]
```diff
@@ -75,31 +75,45 @@ def create_point(x: torch.Tensor, log_q_fn: LogProbFunc, log_p_fn: LogProbFunc,
...
+    if not p_target:
+        assert alpha is not None, "Must specify alpha if AIS target is not p."
     with torch.no_grad():
         # No grad as we don't backprop through this.
-        if p_sq_over_q_target:
-            return (1 - 2*beta) * x.log_q + 2*beta*x.log_p
+        if not p_target:
+            # Use minimum variance importance sampling distribution for alpha-divergence.
+            # AIS target: g = p^\alpha q^(1-\alpha)
+            return ((1-beta) + beta*(1-alpha)) * x.log_q + beta*alpha*x.log_p
```
Comment (lollcat) [inline fab/sampling_methods/base.py:94]: Highlighting: critical change for this MR

File: fab/train_with_prioritised_buffer.py [hunk 3]
```diff
@@ -150,7 +154,7 @@ def run(self,
                 self.optimizer.zero_grad()
                 log_q_x = self.model.flow.log_prob(x)
                 # adjustment to account for change to theta since sample was last added/adjusted
-                log_w_adjust = log_q_old - log_q_x.detach()
+                log_w_adjust = (1-self.alpha) * (log_q_x.detach() - log_q_old)
```
Comment (lollcat) [inline fab/train_with_prioritised_buffer.py:157]: Highlighting: critical change for this MR

File: fab/core.py [hunk 4]
```diff
@@ -85,22 +94,41 @@ def loss(self, args) -> torch.Tensor:
...
+        else:
+            self.annealed_importance_sampler.p_target = False
+            self.annealed_importance_sampler.transition_operator.p_target = False
+
+    def fab_alpha_div_inner(self, point: Point, log_w_ais: torch.Tensor) -> \
+            torch.Tensor:
+        """Compute FAB loss based off points and importance weights from AIS targetting
+        p^\alpha/q^{\alpha-1}.
+        """
+        return - np.sign(self.alpha) * torch.mean(torch.softmax(log_w_ais, dim=-1) * point.log_q)
```
Comment (lollcat) [inline fab/core.py:121]: critical change for this MR

File: fab/trainable_distributions/defensive_mixture.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
+from typing import Optional, Tuple
```
Comment (lollcat) [inline fab/trainable_distributions/defensive_mixture.py:1]: I added this earlier on my branch as I think it may improve things, but currently it is not used anywhere and I need to still do more thorough testing of it.
